File: jvm/src/test/java/io/github/apexdevtools/apexparser/ApexAnnotationTest.java
/*
 Copyright (c) 2026 Kevin Jones, All rights reserved.
 Redistribution and use in source and binary forms, with or without
 modification, are permitted provided that the following conditions
 are met:
 1. Redistributions of source code must retain the above copyright
    notice, this list of conditions and the following disclaimer.
 2. Redistributions in binary form must reproduce the above copyright
    notice, this list of conditions and the following disclaimer in the
    documentation and/or other materials provided with the distribution.
 3. The name of the author may not be used to endorse or promote products
    derived from this software without specific prior written permission.
 */
package io.github.apexdevtools.apexparser;

import static io.github.apexdevtools.apexparser.SyntaxErrorCounter.createParser;
import static org.junit.jupiter.api.Assertions.*;

import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Test;

public class ApexAnnotationTest {

  /* Parse an annotation as a class modifier. compilationUnit ends in EOF, so the whole annotation
     has to be consumed for this to report no errors. */
  private ApexParser.AnnotationContext parseTypeAnnotation(String annotation) {
    Map.Entry<ApexParser, SyntaxErrorCounter> parserAndCounter = createParser(
      annotation + " public class Hello {}"
    );
    ApexParser.CompilationUnitContext context = parserAndCounter
      .getKey()
      .compilationUnit();
    assertEquals(0, parserAndCounter.getValue().getNumErrors());
    return context.typeDeclaration().modifier(0).annotation();
  }

  private int typeAnnotationErrors(String annotation) {
    Map.Entry<ApexParser, SyntaxErrorCounter> parserAndCounter = createParser(
      annotation + " public class Hello {}"
    );
    parserAndCounter.getKey().compilationUnit();
    return parserAndCounter.getValue().getNumErrors();
  }

  private int memberAnnotationErrors(String annotation) {
    Map.Entry<ApexParser, SyntaxErrorCounter> parserAndCounter = createParser(
      "public class Hello { " + annotation + " public void func() {} }"
    );
    parserAndCounter.getKey().compilationUnit();
    return parserAndCounter.getValue().getNumErrors();
  }

  private List<ApexParser.ElementValuePairContext> pairs(String annotation) {
    ApexParser.ElementValuePairsContext pairs = parseTypeAnnotation(
      annotation
    ).elementValuePairs();
    assertNotNull(pairs);
    return pairs.elementValuePair();
  }

  private String valueOf(ApexParser.ElementValuePairContext pair) {
    return pair.elementValue().literal().getText();
  }

  @Test
  void testNoParameters() {
    ApexParser.AnnotationContext context = parseTypeAnnotation("@AuraEnabled");
    assertEquals("AuraEnabled", context.id().getText());
    assertNull(context.LPAREN());
    assertNull(context.elementValuePairs());
    assertNull(context.elementValue());
  }

  @Test
  void testEmptyParameters() {
    ApexParser.AnnotationContext context = parseTypeAnnotation(
      "@AuraEnabled()"
    );
    assertEquals("AuraEnabled", context.id().getText());
    assertNotNull(context.LPAREN());
    assertNull(context.elementValuePairs());
    assertNull(context.elementValue());
  }

  @Test
  void testSingleUnnamedValue() {
    ApexParser.AnnotationContext context = parseTypeAnnotation(
      "@SuppressWarnings('PMD')"
    );
    assertEquals("SuppressWarnings", context.id().getText());
    assertNull(context.elementValuePairs());
    assertNotNull(context.elementValue());
    assertEquals("'PMD'", context.elementValue().literal().getText());
  }

  @Test
  void testSingleNamedValue() {
    List<ApexParser.ElementValuePairContext> pairs = pairs(
      "@SuppressWarnings(value='PMD')"
    );
    assertEquals(1, pairs.size());
    assertEquals("value", pairs.get(0).id().getText());
    assertEquals("'PMD'", valueOf(pairs.get(0)));
  }

  @Test
  void testSpaceSeparatedPairs() {
    List<ApexParser.ElementValuePairContext> pairs = pairs(
      "@IsTest(SeeAllData=true IsParallel=false)"
    );
    assertEquals(2, pairs.size());
    assertEquals("SeeAllData", pairs.get(0).id().getText());
    assertEquals("true", valueOf(pairs.get(0)));
    assertEquals("IsParallel", pairs.get(1).id().getText());
    assertEquals("false", valueOf(pairs.get(1)));
  }

  /* The platform separates parameters by whitespace alone and rejects the comma form. It is
     accepted here deliberately, see the comment on elementValuePairs in BaseApexParser.g4. */
  @Test
  void testCommaSeparatedPairsAreAccepted() {
    List<ApexParser.ElementValuePairContext> pairs = pairs(
      "@IsTest(SeeAllData=true, IsParallel=false)"
    );
    assertEquals(2, pairs.size());
    assertEquals("SeeAllData", pairs.get(0).id().getText());
    assertEquals("IsParallel", pairs.get(1).id().getText());
  }

  @Test
  void testCommaSeparatedPairsAreAcceptedOnAMember() {
    assertEquals(
      0,
      memberAnnotationErrors("@IsTest(SeeAllData=true, IsParallel=false)")
    );
  }

  @Test
  void testTrailingCommaIsRejected() {
    assertEquals(1, typeAnnotationErrors("@IsTest(SeeAllData=true,)"));
  }

  @Test
  void testQuotedValueContainingCommas() {
    ApexParser.AnnotationContext context = parseTypeAnnotation(
      "@SuppressWarnings('PMD,Unused')"
    );
    assertEquals("'PMD,Unused'", context.elementValue().literal().getText());
  }

  @Test
  void testQuotedValueContainingSpaces() {
    List<ApexParser.ElementValuePairContext> pairs = pairs(
      "@InvocableMethod(label='Do a thing' category='Some, category')"
    );
    assertEquals(2, pairs.size());
    assertEquals("'Do a thing'", valueOf(pairs.get(0)));
    assertEquals("'Some, category'", valueOf(pairs.get(1)));
  }

  @Test
  void testStringLiteralValue() {
    assertNotNull(
      pairs("@InvocableMethod(label='x')")
        .get(0)
        .elementValue()
        .literal()
        .StringLiteral()
    );
  }

  @Test
  void testMultilineStringLiteralValue() {
    assertNotNull(
      pairs("@InvocableMethod(label='''\nx''')")
        .get(0)
        .elementValue()
        .literal()
        .MultilineStringLiteral()
    );
  }

  @Test
  void testBooleanLiteralValue() {
    assertNotNull(
      pairs("@AuraEnabled(cacheable=true)")
        .get(0)
        .elementValue()
        .literal()
        .BooleanLiteral()
    );
  }

  @Test
  void testIntegerLiteralValue() {
    assertNotNull(
      pairs("@InvocableMethod(label=42)")
        .get(0)
        .elementValue()
        .literal()
        .IntegerLiteral()
    );
  }

  @Test
  void testLongLiteralValue() {
    assertNotNull(
      pairs("@InvocableMethod(label=42L)")
        .get(0)
        .elementValue()
        .literal()
        .LongLiteral()
    );
  }

  @Test
  void testNumberLiteralValue() {
    assertNotNull(
      pairs("@InvocableMethod(label=4.2)")
        .get(0)
        .elementValue()
        .literal()
        .NumberLiteral()
    );
  }

  @Test
  void testNullLiteralValue() {
    assertNotNull(
      pairs("@AuraEnabled(cacheable=null)")
        .get(0)
        .elementValue()
        .literal()
        .NULL()
    );
  }

  /* Values the platform rejects still parse, so that apex-ls can diagnose them precisely rather
     than the file being lost to a syntax error. See apex-ls#326. */
  @Test
  void testTypeMismatchedValueStillParses() {
    assertEquals(0, typeAnnotationErrors("@AuraEnabled(cacheable=0)"));
    assertEquals(0, typeAnnotationErrors("@AuraEnabled(cacheable=null)"));
    assertEquals(0, typeAnnotationErrors("@InvocableMethod(label=42)"));
  }

  @Test
  void testQualifiedNameIsRejected() {
    assertTrue(typeAnnotationErrors("@Schema.AuraEnabled") > 0);
    assertTrue(memberAnnotationErrors("@Schema.AuraEnabled") > 0);
  }

  @Test
  void testArrayInitializerValueIsRejected() {
    assertTrue(typeAnnotationErrors("@InvocableMethod(label={'a','b'})") > 0);
    assertTrue(memberAnnotationErrors("@InvocableMethod(label={'a','b'})") > 0);
  }

  @Test
  void testNestedAnnotationValueIsRejected() {
    assertTrue(typeAnnotationErrors("@InvocableMethod(label=@IsTest)") > 0);
    assertTrue(memberAnnotationErrors("@InvocableMethod(label=@IsTest)") > 0);
  }

  @Test
  void testBareIdentifierValueIsRejected() {
    assertTrue(typeAnnotationErrors("@AuraEnabled(cacheable=foo)") > 0);
    assertTrue(memberAnnotationErrors("@AuraEnabled(cacheable=foo)") > 0);
  }

  @Test
  void testBareIdentifierUnnamedValueIsRejected() {
    assertTrue(typeAnnotationErrors("@SuppressWarnings(PMD)") > 0);
    assertTrue(memberAnnotationErrors("@SuppressWarnings(PMD)") > 0);
  }

  @Test
  void testExpressionValueIsRejected() {
    assertTrue(typeAnnotationErrors("@InvocableMethod(label='a' + 'b')") > 0);
    assertTrue(memberAnnotationErrors("@InvocableMethod(label='a' + 'b')") > 0);
  }
}

File: npm/test/ApexAnnotationTest.ts
/*
 Copyright (c) 2026 Kevin Jones, All rights reserved.
 Redistribution and use in source and binary forms, with or without
 modification, are permitted provided that the following conditions
 are met:
 1. Redistributions of source code must retain the above copyright
    notice, this list of conditions and the following disclaimer.
 2. Redistributions in binary form must reproduce the above copyright
    notice, this list of conditions and the following disclaimer in the
    documentation and/or other materials provided with the distribution.
 3. The name of the author may not be used to endorse or promote products
    derived from this software without specific prior written permission.
 */
import {
  AnnotationContext,
  ElementValuePairContext,
} from "../src/antlr/ApexParser.js";
import { createParser } from "./SyntaxErrorCounter.js";

/* Parse an annotation as a class modifier. compilationUnit ends in EOF, so the whole annotation
   has to be consumed for this to report no errors. */
function parseTypeAnnotation(annotation: string): AnnotationContext {
  const [parser, errorCounter] = createParser(
    `${annotation} public class Hello {}`
  );
  const context = parser.compilationUnit();

  expect(errorCounter.getNumErrors()).toEqual(0);
  return context.typeDeclaration().modifier(0).annotation();
}

function typeAnnotationErrors(annotation: string): number {
  const [parser, errorCounter] = createParser(
    `${annotation} public class Hello {}`
  );
  parser.compilationUnit();
  return errorCounter.getNumErrors();
}

function memberAnnotationErrors(annotation: string): number {
  const [parser, errorCounter] = createParser(
    `public class Hello { ${annotation} public void func() {} }`
  );
  parser.compilationUnit();
  return errorCounter.getNumErrors();
}

function pairs(annotation: string): ElementValuePairContext[] {
  const pairs = parseTypeAnnotation(annotation).elementValuePairs();
  expect(pairs).toBeTruthy();
  return pairs.elementValuePair_list();
}

function valueOf(pair: ElementValuePairContext): string {
  return pair.elementValue().literal().getText();
}

test("No parameters", () => {
  const context = parseTypeAnnotation("@AuraEnabled");

  expect(context.id().getText()).toBe("AuraEnabled");
  expect(context.LPAREN()).toBeNull();
  expect(context.elementValuePairs()).toBeNull();
  expect(context.elementValue()).toBeNull();
});

test("Empty parameters", () => {
  const context = parseTypeAnnotation("@AuraEnabled()");

  expect(context.id().getText()).toBe("AuraEnabled");
  expect(context.LPAREN()).toBeTruthy();
  expect(context.elementValuePairs()).toBeNull();
  expect(context.elementValue()).toBeNull();
});

test("Single unnamed value", () => {
  const context = parseTypeAnnotation("@SuppressWarnings('PMD')");

  expect(context.id().getText()).toBe("SuppressWarnings");
  expect(context.elementValuePairs()).toBeNull();
  expect(context.elementValue().literal().getText()).toBe("'PMD'");
});

test("Single named value", () => {
  const params = pairs("@SuppressWarnings(value='PMD')");

  expect(params.length).toBe(1);
  expect(params[0].id().getText()).toBe("value");
  expect(valueOf(params[0])).toBe("'PMD'");
});

test("Space separated pairs", () => {
  const params = pairs("@IsTest(SeeAllData=true IsParallel=false)");

  expect(params.length).toBe(2);
  expect(params[0].id().getText()).toBe("SeeAllData");
  expect(valueOf(params[0])).toBe("true");
  expect(params[1].id().getText()).toBe("IsParallel");
  expect(valueOf(params[1])).toBe("false");
});

// The platform separates parameters by whitespace alone and rejects the comma form. It is
// accepted here deliberately, see the comment on elementValuePairs in BaseApexParser.g4.
test("Comma separated pairs are accepted", () => {
  const params = pairs("@IsTest(SeeAllData=true, IsParallel=false)");

  expect(params.length).toBe(2);
  expect(params[0].id().getText()).toBe("SeeAllData");
  expect(params[1].id().getText()).toBe("IsParallel");
});

test("Comma separated pairs are accepted on a member", () => {
  expect(
    memberAnnotationErrors("@IsTest(SeeAllData=true, IsParallel=false)")
  ).toEqual(0);
});

test("Trailing comma is rejected", () => {
  expect(typeAnnotationErrors("@IsTest(SeeAllData=true,)")).toEqual(1);
});

test("Quoted value containing commas", () => {
  const context = parseTypeAnnotation("@SuppressWarnings('PMD,Unused')");

  expect(context.elementValue().literal().getText()).toBe("'PMD,Unused'");
});

test("Quoted value containing spaces", () => {
  const params = pairs(
    "@InvocableMethod(label='Do a thing' category='Some, category')"
  );

  expect(params.length).toBe(2);
  expect(valueOf(params[0])).toBe("'Do a thing'");
  expect(valueOf(params[1])).toBe("'Some, category'");
});

test("String literal value", () => {
  const literal = pairs("@InvocableMethod(label='x')")[0]
    .elementValue()
    .literal();

  expect(literal.StringLiteral()).toBeTruthy();
});

test("Multiline string literal value", () => {
  const literal = pairs("@InvocableMethod(label='''\nx''')")[0]
    .elementValue()
    .literal();

  expect(literal.MultilineStringLiteral()).toBeTruthy();
});

test("Boolean literal value", () => {
  const literal = pairs("@AuraEnabled(cacheable=true)")[0]
    .elementValue()
    .literal();

  expect(literal.BooleanLiteral()).toBeTruthy();
});

test("Integer literal value", () => {
  const literal = pairs("@InvocableMethod(label=42)")[0]
    .elementValue()
    .literal();

  expect(literal.IntegerLiteral()).toBeTruthy();
});

test("Long literal value", () => {
  const literal = pairs("@InvocableMethod(label=42L)")[0]
    .elementValue()
    .literal();

  expect(literal.LongLiteral()).toBeTruthy();
});

test("Number literal value", () => {
  const literal = pairs("@InvocableMethod(label=4.2)")[0]
    .elementValue()
    .literal();

  expect(literal.NumberLiteral()).toBeTruthy();
});

test("Null literal value", () => {
  const literal = pairs("@AuraEnabled(cacheable=null)")[0]
    .elementValue()
    .literal();

  expect(literal.NULL()).toBeTruthy();
});

// Values the platform rejects still parse, so that apex-ls can diagnose them precisely rather
// than the file being lost to a syntax error. See apex-ls#326.
test("Type mismatched value still parses", () => {
  expect(typeAnnotationErrors("@AuraEnabled(cacheable=0)")).toEqual(0);
  expect(typeAnnotationErrors("@AuraEnabled(cacheable=null)")).toEqual(0);
  expect(typeAnnotationErrors("@InvocableMethod(label=42)")).toEqual(0);
});

test("Qualified name is rejected", () => {
  expect(typeAnnotationErrors("@Schema.AuraEnabled")).toBeGreaterThan(0);
  expect(memberAnnotationErrors("@Schema.AuraEnabled")).toBeGreaterThan(0);
});

test("Array initializer value is rejected", () => {
  expect(
    typeAnnotationErrors("@InvocableMethod(label={'a','b'})")
  ).toBeGreaterThan(0);
  expect(
    memberAnnotationErrors("@InvocableMethod(label={'a','b'})")
  ).toBeGreaterThan(0);
});

test("Nested annotation value is rejected", () => {
  expect(
    typeAnnotationErrors("@InvocableMethod(label=@IsTest)")
  ).toBeGreaterThan(0);
  expect(
    memberAnnotationErrors("@InvocableMethod(label=@IsTest)")
  ).toBeGreaterThan(0);
});

test("Bare identifier value is rejected", () => {
  expect(typeAnnotationErrors("@AuraEnabled(cacheable=foo)")).toBeGreaterThan(
    0
  );
  expect(memberAnnotationErrors("@AuraEnabled(cacheable=foo)")).toBeGreaterThan(
    0
  );
});

test("Bare identifier unnamed value is rejected", () => {
  expect(typeAnnotationErrors("@SuppressWarnings(PMD)")).toBeGreaterThan(0);
  expect(memberAnnotationErrors("@SuppressWarnings(PMD)")).toBeGreaterThan(0);
});

test("Expression value is rejected", () => {
  expect(
    typeAnnotationErrors("@InvocableMethod(label='a' + 'b')")
  ).toBeGreaterThan(0);
  expect(
    memberAnnotationErrors("@InvocableMethod(label='a' + 'b')")
  ).toBeGreaterThan(0);
});
